lib/elixir/pages/Guards.md

```diff
@@ -0,0 +1,226 @@
+
+```elixir
+def long_list?([]),               do: false
+def long_list?([_]),              do: false
+def long_list?([_, _]),           do: false
+def long_list?([_, _, _]),        do: false
+def long_list?([_, _, _, _ | _]), do: true
+```
+
+This is verbose and not scalable with the length of "long lists". Guards solve this by letting us performs additional and more complex checks than just pattern matching, allowing us to simplify clauses or even express something that is not possible to express through pattern matching.
+
+## Where guards can be used
+
+In the example above, we show how guards can be used in function clauses. In total, there are **five** constructs that allow guards:
+
+  * function clauses:
+
+  ```elixir
+  def foo(x) when is_integer(x), do: x
+  def foo(x) when is_float(x), do: round(x)
+  ```
+
+  * `case` expressions:
+
+  ```elixir
+  case x do
+    1 -> :one
+    2 -> :two
+    n when is_integer(n) and n > 2 -> :larger_than_two
+  end
+  ```
+
+  * anonymous functions (`fn`s):
+
+  ```elixir
+  larger_than_two? = fn
+    n when is_integer(n) and n > 2 -> true
+    n when is_integer(n) -> false
+  end
+  ```
+
+  * `for` comprehensions:
+
+  ```elixir
+  for n when n > 0 <- my_integers, do: n * 2
+  ```
+
+  * `with` expressions:
+
+  ```elixir
+  with {:ok, n} when is_integer(n) <- my_operation(), do: {:ok, n * 2}
+  ```
+
+Guards cannot be used in every place where pattern match is allowed. For example, guards aren't allowed with the pattern matching operator (`=`):
+
+```elixir
+# This is not allowed (and results in a syntax error)
+[head | tail] when is_integer(head) = my_list
+```
+
+## Failing guards
+
+An important thing to keep in mind is that errors in guards do not result in a runtime error, but in the erroring guard fail. For example, the `length/1` function only works with lists, and if we use it on anything else it fails:
+
+```elixir
+iex> length("hello")
+** (ArgumentError) argument error
+```
+
+However, when used in guards, it simply makes the corresponding clause fail:
+
+```elixir
+iex> case "hello" do
+...>   something when length(something) > 0 ->
+...>     :length_worked
+...>   _anything_else ->
+...>     :length_failed
+...> end
+:length_failed
+```
+
+In many cases, we can take advantage of this: in the code above, for example, we can use `length/1` to both check that the given thing is a list *and* check some properties of its length (instead of using `is_list(something) and length(something) > 0`).
+
+## Expressions in guard clauses
+
+Not all expressions are allowed in guard clauses, but only a handful of them. This is a deliberate choice: only a predefined set of side-effect-free functions are allowed. This way, Elixir (and Erlang) can make sure that nothing bad happens while executing guards and no mutations happen anywhere. This behaviour is also coherent with pattern match, which is a naturally a side-effect-free operation.
```

Comment:
Maybe worth mentioning that keeping the clauses to a known set also allows the compiler to optimize the code.